### PR TITLE
adds mapWithIndex function.

### PR DESCRIPTION
**Pull request by kRITZCREEK**
This is the simplest implementation I could think of. Please check if there are performance troubles.

### CODE REVIEW

File: src/Data/List.purs [hunk 1]
```diff
@@ -428,6 +429,14 @@ mapMaybe f = go Nil
 catMaybes :: forall a. List (Maybe a) -> List a
 catMaybes = mapMaybe id
 
+
+-- | Apply a function to each element and its index in a list starting at 0.
+mapWithIndex :: forall a b. (a -> Int -> b) -> List a -> List b
+mapWithIndex f = go 0
+  where
+  go _ Nil = Nil
+  go n (Cons a as) = Cons (f a n) (go (n+1) as)
```
Comment (paf31) [inline src/Data/List.purs]: If we could make this tail recursive, that would be nice.

**Comment by paf31** [comment]
:+1: LGTM, thanks!

**Comment by garyb** [comment]
:+1:

**Comment by garyb** [comment]
Thanks!

**Pull request merged**